### DISCUSSION TITLE
How to use Rust crates in applications?

### DISCUSSION THREAD

**Discussion by george-hopkins** [unikraft/unikraft#400] [Q&A] [answered]
With the initial support for Rust thanks to #241, I played around with it a bit and implemented a small driver for virtio-rng. As the no-std ecosystem is growing steadily, it would be nice to integrate existing crates. Is there a recommended way to do so?

I got it working by compiling an external crate with `--emit obj` and adding all object files (of the crate and the dependencies) to the Makefile. However, this approach is just a half-baked solution and breaks if I recompile the crate without cleaning the target directory first.

**Comment by vladandrew** [comment, marked as answer]
Hi,

Yes, @TrueDoctor has been doing some work on this and actually got it working. There is a pull request[1] that will be soon updated along with the PoC from @TrueDoctor[2]. We are planning to upstream his work ASAP. We're planning on having an external library called lib-cargo that would enable cargo support.

[1] https://github.com/unikraft/unikraft/pull/291
[2] https://github.com/unikraft/unikraft/pulls?q=is%3Apr+rust+is%3Aclosed

**Comment by razvand** [comment]
@george-hopkins, since we are already (viciously) using [the Unikraft Discord server](https://bit.ly/UnikraftDiscord) for discussions, meetings, brainstorming and jokes (mostly bad, we do what we can), we will close the Discussions tab. We will do this on Friday, March 31, 2023.

Please move any discussions or questions on [the Unikraft Discord server](https://bit.ly/UnikraftDiscord). See you all there! :handshake:
